### PR TITLE
Flink: Test both "new" Flink Avro planned reader and "deprecated" Avro reader

### PR DESCRIPTION
**Pull request by jbonofre**
Following #11386, this PR adds test coverage on both the "new" `FlinkAvroPlannedReader` and the "deprecated" `FlinkAvroReader`.

### CODE REVIEW

**Comment by jbonofre** [comment]
@RussellSpitzer as requested by @pvary and @Fokko this PR add tests for both the "old" and "new" readers.

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java [hunk 1]
```diff
@@ -65,12 +63,13 @@ public class TestFlinkAvroReaderWriter extends DataTest {
...
-  protected void writeAndValidate(Schema schema) throws IOException {
-    List<Record> expectedRecords = RandomGenericData.generate(schema, NUM_RECORDS, 1991L);
-    writeAndValidate(schema, expectedRecords, NUM_RECORDS);
-  }
+  protected abstract void writeAndValidate(Schema schema) throws IOException;
 
-  private void writeAndValidate(Schema schema, List<Record> expectedRecords, int numRecord)
+  void writeAndValidate(
+      Schema schema,
+      List<org.apache.iceberg.data.Record> expectedRecords,
```
Comment (pvary) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java:70]: Does it need to be fully qualified?

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java [hunk 2]
```diff
@@ -80,20 +79,27 @@ private void writeAndValidate(Schema schema, List<Record> expectedRecords, int n
...
-            .createResolvingReader(FlinkPlannedAvroReader::create)
-            .build()) {
-      Iterator<Record> expected = expectedRecords.iterator();
+            .createReaderFunc(FlinkAvroReader::new);
+    if (useAvroPlannedReader) {
+      builder =
+          Avro.read(Files.localInput(recordsFile))
+              .project(schema)
+              .createResolvingReader(FlinkPlannedAvroReader::create);
+    }
```
Comment (pvary) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java:99]: Maybe put this in an abstract method and use it inside the `writeAndValidate`?
Something like:
```
abstract Avro.ReadBuilder builder() {
```

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestRowProjection.java [hunk 3]
```diff
@@ -61,52 +63,63 @@ private RowData writeAndRead(String desc, Schema writeSchema, Schema readSchema,
...
+    }
+    Iterable<RowData> records = builder.build();
 
     return Iterables.getOnlyElement(records);
   }
 
-  @Test
-  public void testFullProjection() throws Exception {
+  @ParameterizedTest
+  @ValueSource(booleans = {false, true})
```
Comment (pvary) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestRowProjection.java:82]: Do we add this to all of the tests? Maybe use `@TestTemplate` for this

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java (hunk 1 above)
Comment (jbonofre) [reply flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java:70]: Afair, the reason is to avoid confusion with another `Record`. Let me check.

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java (hunk 2 above)
Comment (jbonofre) [reply flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java:99]: Good idea. Let me do that. Thanks !

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestRowProjection.java (hunk 3 above)
Comment (jbonofre) [reply flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestRowProjection.java:82]: Let me check, but I think `@TestTemplate` is an option.

Comment (jbonofre) [reply flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestRowProjection.java:82]: Actually, after reviewing this part, I'm not convinced `@TestTemplate` is much better: we would need to add `@TestTemplate` and `@ExtendWith()` annotations. So, still two annotations to add + the context provider.

@pvary thoughts ?

**Comment by jbonofre** [comment]
@Fokko @pvary @RussellSpitzer I updated the tests according to Peter's comment. I'm just not convinced that using `@TestTemplate` is much better than `@ParameterizedTest` 😄

**Comment by jbonofre** [comment]
@nastra if you have time to take a look, thanks ! 😄

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestRowProjection.java (hunk 3 above)
Comment (nastra) [reply flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestRowProjection.java:82]: `@TestTemplate` is for parameterization at the class level, which seems exactly what's being done here. So rather than adding `@ValueSource` to each single method it makes sense to do this at the class level and annotate all tests with `@TestTemplate`

Comment (jbonofre) [reply flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestRowProjection.java:82]: OK, let me do it.

Comment (jbonofre) [reply flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestRowProjection.java:82]: @nastra by the way, I was confused about the expectation here because `@TestTemplate` is not necessary at class level, it can be at method level too. Actually both `@ParameterizedTest` and `@RepeatedTest` are based on `@TestTemplate` (they use custom invocation context provider). That's why I was confused (as `@ParameterizedTest` is actually a "specialized" `@TestTemplate`) 😄

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java [hunk 4]
```diff
@@ -80,20 +77,18 @@ private void writeAndValidate(Schema schema, List<Record> expectedRecords, int n
 
     // Write the expected records into AVRO file, then read them into RowData and assert with the
     // expected Record list.
-    try (FileAppender<Record> writer =
+    try (FileAppender<org.apache.iceberg.data.Record> writer =
```
Comment (pvary) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java:80]: Do we really need the fully qualified name here? I'm reviewing on my phone, so I might miss something, but I was not able to find the reason

Comment (jbonofre) [reply flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java:80]: No, we can use the short name. I will update.

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestRowProjection.java (hunk 3 above)
Comment (jbonofre) [reply flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestRowProjection.java]: I update the test to use `@TestTemplate`.

**Comment by jbonofre** [comment]
@nastra @pvary @Fokko I updated the tests according to your comments. If you can take a new look, thanks !

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestFlinkAvroPlannedReaderWriter.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import org.apache.iceberg.Files;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.avro.Avro;
+import org.apache.iceberg.data.RandomGenericData;
+import org.apache.iceberg.data.Record;
+
+public class TestFlinkAvroPlannedReaderWriter extends AbstractTestFlinkAvroReaderWriter {
+
+  @Override
+  protected void writeAndValidate(Schema schema) throws IOException {
```
Comment (nastra) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestFlinkAvroPlannedReaderWriter.java:33]: not sure why both subclasses need to override this method from the super class

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java [hunk 6]
```diff
@@ -65,12 +63,11 @@ public class TestFlinkAvroReaderWriter extends DataTest {
           Types.NestedField.optional(9, "decimal", Types.DecimalType.of(4, 2)));
 
   @Override
-  protected void writeAndValidate(Schema schema) throws IOException {
-    List<Record> expectedRecords = RandomGenericData.generate(schema, NUM_RECORDS, 1991L);
-    writeAndValidate(schema, expectedRecords, NUM_RECORDS);
-  }
+  protected abstract void writeAndValidate(Schema schema) throws IOException;
```
Comment (nastra) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java:66]: is there a particular reason why this is being made abstract now? It seems both subclasses override and do the same thing as this method was doing previously

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java [hunk 7]
```diff
@@ -154,9 +149,7 @@ private Record recordNumType(
     return record;
   }
 
-  @Test
-  public void testNumericTypes() throws IOException {
-
+  void testNumericTypes() throws IOException {
```
Comment (nastra) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java:152]: it seems this became a simple method that isn't being tested anymore?

**Comment by jbonofre** [comment]
@nastra sorry, I forgot to push some changes, it should be better now.

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestRowProjection.java [hunk 8]
```diff
@@ -61,16 +76,22 @@ private RowData writeAndRead(String desc, Schema writeSchema, Schema readSchema,
...
-            .createResolvingReader(FlinkPlannedAvroReader::create)
-            .build();
+            .createReaderFunc(FlinkAvroReader::new);
+    if (useAvroPlannedReader) {
+      builder =
+          Avro.read(Files.localInput(file))
+              .project(readSchema)
+              .createResolvingReader(FlinkPlannedAvroReader::create);
+    }
+    Iterable<RowData> records = builder.build();
```
Comment (nastra) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestRowProjection.java:89]: nit: newline after }

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java [hunk 9]
```diff
@@ -64,12 +64,13 @@ public class TestFlinkAvroReaderWriter extends DataTest {
           Types.NestedField.optional(8, "bigint", Types.LongType.get()),
           Types.NestedField.optional(9, "decimal", Types.DecimalType.of(4, 2)));
 
-  @Override
```
Comment (nastra) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java:67]: unrelated change

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java [hunk 10]
```diff
@@ -48,11 +48,11 @@
 import org.apache.iceberg.util.DateTimeUtil;
 import org.junit.jupiter.api.Test;
 
-public class TestFlinkAvroReaderWriter extends DataTest {
+public abstract class AbstractTestFlinkAvroReaderWriter extends DataTest {
 
-  private static final int NUM_RECORDS = 100;
+  static final int NUM_RECORDS = 100;
```
Comment (nastra) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java:53]: this and the ones below are both unrelated changes I believe

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java [hunk 11]
```diff
@@ -88,11 +89,9 @@ private void writeAndValidate(Schema schema, List<Record> expectedRecords, int n
...
     }
 
-    try (CloseableIterable<RowData> reader =
-        Avro.read(Files.localInput(recordsFile))
-            .project(schema)
-            .createResolvingReader(FlinkPlannedAvroReader::create)
-            .build()) {
+    Avro.ReadBuilder builder = createAvroReadBuilder(recordsFile, schema);
+
+    try (CloseableIterable<RowData> reader = builder.build()) {
```
Comment (nastra) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java:94]:
```suggestion
    try (CloseableIterable<RowData> reader = createAvroReadBuilder(recordsFile, schema).build()) {
```

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java [hunk 12]
```diff
@@ -156,7 +154,7 @@ private Record recordNumType(
 
   @Test
   public void testNumericTypes() throws IOException {
-
+
```
Comment (nastra) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/AbstractTestFlinkAvroReaderWriter.java:157]: removing that line in your earlier commit was fine imo, but this is adding a tab ;)

**Review by nastra — APPROVED** [review]

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestFlinkAvroDeprecatedReaderWriter.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+ * under the License.
+ */
+package org.apache.iceberg.flink.data;
+
+import java.io.File;
+import org.apache.iceberg.Files;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.avro.Avro;
+
+public class TestFlinkAvroDeprecatedReaderWriter extends AbstractTestFlinkAvroReaderWriter {
```
Comment (pvary) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/data/TestFlinkAvroDeprecatedReaderWriter.java:26]: Nit: could we add a deprecated annotation and a comment:
```
/**
 * @deprecated should be removed in 1.8.0; along with FlinkAvroReader.
 */
@Deprecated
```

**Review by pvary — APPROVED** [review]

**Comment by jbonofre** [comment]
@RussellSpitzer do you mind to merge this one ? Thanks !

**Comment by pvary** [comment]
@jbonofre: If the deprecated message is there, then I can merge.. no problem

**Comment by jbonofre** [comment]
@pvary I may have missed your comment (about deprecated message) 😀

**Comment by jbonofre** [comment]
@pvary I added `@Deprecated` to the test.

**Comment by pvary** [comment]
Thanks @jbonofre for the PR and @nastra for the review.

@jbonofre: Please port the changes to Flink 1.19/1.18 too.
Thanks,
Peter
